### PR TITLE
[SPARK-4452][Core]Shuffle data structures can starve others on the same thread for memory

### PR DESCRIPTION
**Pull request by lianhuiwang**
## What changes were proposed in this pull request?

In #9241 It implemented a mechanism to call spill() on those SQL operators that support spilling if there is not enough memory for execution. 
But ExternalSorter and AppendOnlyMap in Spark core are not worked. So this PR make them benefit from #9241. Now when there is not enough memory for execution, it can get memory by spilling ExternalSorter and AppendOnlyMap in Spark core.
## How was this patch tested?

add two unit tests for it.

### CODE REVIEW

**Comment by lianhuiwang** [comment]
test this please

**Comment by squito** [comment]
Hi @lianhuiwang thanks for submitting this patch.

I just have a really high-level question first.  If I understand how this works correctly, the idea is that:
1) as data is added to `Spillable`s, they work exactly the same as before at first, so after all records have been inserted, they've still got a bunch of data in memory.
2) When an iterator is requested, as before its still an iterator that merges data has been spilled to disk, and data that is still-in-memory
3) If the `TaskMemoryManager` requests more memory while that iterator is in flight, then the `Spillable`s look at the position of the current iterator over the in-memory data, and spill only the _remaining_ data to disk
4) The `Spillable` then free the memory to the `TaskMemoryManager`, and have the in-flight iterator switch to using the new spill file.

Is this a correct understanding?  If so, this seems to hinge on one key assumption: that the `Spillable`s are never iterated over multiple times.  If they were iterated multiple times, then the second iteration would be incorrect -- some of the initial data in the in-memory structure would be lost on the second iteration.

I think this assumption is sound -- it is implied by "destructive"SortedIterator in the internals, though I think the actual wrapping `Spillable`s might allow multiple iteration now.  I've been trying to think of a case where this assumption would be violated but cant' come up with anything.  (If the same shuffle data is read multiple times in different tasks, the work on the shuffle-read side is simply repeated each time, I'm pretty sure there isn't any sharing).  But nonetheless if that is the case, I think this deserves both a lot of comments explaining how this works, assertions which make sure this assumption is not violated, and a number of tests.

FWIW, I started down the path of writing something similar with*out* that assumption -- when a spill was requested on an in-flight iterator, then the *entire* in-memory structure would get spilled to disk, and the in-flight iterator would switch to the spilled data, and advance to the same location in the spilled data that it was on the in-memory data.  This was pretty convoluted, and as I started writing tests I realized there were corner cases that needed work.  So I decided to submit the simpler change instead.  It seems much easier to it your way.    I do have some test which I think I can add as well -- lemme dig those up and send them later today.

File: core/src/main/java/org/apache/spark/memory/MemoryConsumer.java [hunk 1]
```diff
@@ -130,4 +130,22 @@ protected void freePage(MemoryBlock page) {
     used -= page.size();
     taskMemoryManager.freePage(page, this);
   }
+
+  /**
+   * Allocates a heap memory of `size`.
+   */
+  public long allocateHeapExecutionMemory(long size) {
```
Comment (davies) [inline core/src/main/java/org/apache/spark/memory/MemoryConsumer.java]: This function does not actually create any object, I'd like to call it `acquireOnHeapMemory`

File: core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala [hunk 2]
```diff
@@ -235,6 +237,52 @@ private[spark] class ExternalSorter[K, V, C](
...
+    val spillFile = spillMemoryIteratorToDisk(inMemoryIterator)
+    spills.append(spillFile)
+  }
+
+  /**
+   * Force to spilling the current in-memory collection to disk to release memory,
+   * It will be called by TaskMemoryManager when there is not enough memory for the task.
+   */
+  override protected[this] def forceSpill(): Boolean = {
+    if (isShuffleSort) {
```
Comment (davies) [inline core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala]: This could be triggered by a different thread, so it should be thread safe.

File: core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala [hunk 3]
```diff
@@ -689,11 +760,18 @@ private[spark] class ExternalSorter[K, V, C](
...
+    forceSpillFile.foreach(_.file.delete())
+    if (map != null || buffer != null) {
+      map = null // So that the memory can be garbage-collected
+      buffer = null // So that the memory can be garbage-collected
+      releaseMemory()
+    }
+  }
+
+  override def toString(): String = {
+    this.getClass.getName + "@" + java.lang.Integer.toHexString(this.hashCode())
```
Comment (davies) [inline core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala]: Why this?

**Comment by davies** [comment]
@lianhuiwang Thanks for working on this, I think it's in the good direction. Two things left:

1) thread safety. For example, you will have two threads for PythonRDD (same for RRDD), one iterate rows from parent RDD, another iterator rows from PythonRDD/RRDD, the second one could trigger spilling, the spilling happen in second thread, and the first thread could consuming the same iterator. So must make them thread safe. This is the hardest part, you could take the SQL operators as examples.

2) Adding more tests. As @squito suggested, more comments to explain the high level ideas will be good to have.

**Comment by lianhuiwang** [comment]
@davies Thanks. I have added a SpillableIterator that can make consumer and spill thread safe. I think you can take a look at it.

**Comment by lianhuiwang** [comment]
@squito  Yes, I think your understanding is correct. this PR only support that a Spillables will be called once.  The code 'val sort = new Spillable() sort.iterator() sort.iterator()' will be wrong.
for ExternalSorter.iterator, I find that is just called by BlockStoreShuffleReader.Read(). Every Reader will create new ExternalSorter, So i think it's ok for it.
for ExternalAppendOnlyMap, It has two places that use it.They are Aggregator and CoGroupedRDD. we also can find that it create new ExternalAppendOnlyMap before call one  iterator(). So i think it is also ok for it.
The input and output of them are Iterator and Iterator cannot be cached, so it will not affect when rdds are cached.
BTW: this PR has been running for a long time on our many online spark jobs.
If i missed some points, please put forward. It is worth mentioning that

File: core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala [hunk 4]
```diff
@@ -727,4 +779,64 @@ private[spark] class ExternalSorter[K, V, C](
       (elem._1._2, elem._2)
     }
   }
+
+  private[this] class SpillableIterator(var upstream: Iterator[((Int, K), C)])
```
Comment (davies) [inline core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala]: This should be thread safe.

**Comment by lianhuiwang** [comment]
@davies Now i make some variables volatile to avoid thread safe.  Can you take a look? Thanks.

**Comment by davies** [comment]
@lianhuiwang I don't think this is enough.

If it's not easy to make it thread safe, one option could be not do forced spilling if spill() will called from another thread than the one consuming the iterator. At least it's better than before.

**Comment by davies** [comment]
Could you also rebased to master?

**Comment by lianhuiwang** [comment]
@davies  Yes, I have update it using object.lock. I will rebased to master.

**Comment by lianhuiwang** [comment]
test it please.

**Comment by lianhuiwang** [comment]
Jenkins, test this please

**Comment by lianhuiwang** [comment]
@davies Now all tests have been passed. So Could you take a look again? Thanks.

File: core/src/main/scala/org/apache/spark/util/collection/Spillable.scala [hunk 5]
```diff
@@ -100,6 +103,27 @@ private[spark] trait Spillable[C] extends Logging {
...
+    if (trigger != this && taskMemoryManager.getTungstenMemoryMode == MemoryMode.ON_HEAP) {
+      val isSpilled = forceSpill()
+      if (!isSpilled) {
+        0L
+      } else {
+        _elementsRead = 0
+        val freeMemory = myMemoryThreshold - initialMemoryThreshold
+        _memoryBytesSpilled += freeMemory
+        releaseMemory()
+        freeMemory
```
Comment (davies) [inline core/src/main/scala/org/apache/spark/util/collection/Spillable.scala]: We should free memory first, then release memory

**Comment by davies** [comment]
LGTM

**Comment by davies** [comment]
@lianhuiwang Have you run some stress tests with latest change?

**Comment by zzcclp** [comment]
@davies , will this pr be backported to branch-1.6?

**Comment by davies** [comment]
This is a big change, maybe not.

**Comment by zzcclp** [comment]
sorry, I mistakenly deleted my comment.
what a pity, i can only merge this pr manually.

**Comment by lianhuiwang** [comment]
@davies I also have run unit tests with big number N. 
How about add a config "spark.shuffle.spill.reservedMemory"? its default is true. It does not force to spilling when its value is true. Maybe 2.1 version the config can be deleted.

File: core/src/main/scala/org/apache/spark/util/collection/Spillable.scala (hunk 5 above)
Comment (lianhuiwang) [inline core/src/main/scala/org/apache/spark/util/collection/Spillable.scala]: It did collection = null in forceSpill() before releaseMemory().

**Comment by lianhuiwang** [comment]
Jenkins, test this please

**Comment by davies** [comment]
Merging this into master, thanks!

**Comment by lianhuiwang** [comment]
@davies Thanks.
